cmd/ndm_daemonset/probe/mountprobe.go

```diff
@@ -106,3 +148,85 @@ func (mp *mountProbe) FillBlockDeviceDetails(blockDevice *blockdevice.BlockDevic
 		blockDevice.FSInfo.FileSystem = basicMountInfo.FileSystem
 	}
 }
+
+func (mp *mountProbe) setupEpoll() error {
+	ep, err := epoll.New()
+	if err != nil {
+		return err
+	}
+	mp.epoll = &ep
+	return ep.AddWatcher(epoll.Watcher{
+		FileName:   mp.mountsFileName,
+		EventTypes: []epoll.EventType{epoll.EPOLLERR, epoll.EPOLLPRI},
+	})
+}
+
+func (mp *mountProbe) listen() {
+	eventCh, err := mp.epoll.Start()
+	if err != nil {
+		klog.Errorf("error while starting epoll: %v", err)
+		return
+	}
+	defer mp.epoll.Stop()
+	klog.Info("started mount change detection loop")
+	defaultMsg := controller.EventMessage{
+		Action:          string(MountEA),
+		Devices:         nil,
+		AllBlockDevices: true,
+	}
+
+	for range eventCh {
+		// regenerate mounts table and get the changes
+		newMountTable, err := mp.newMountTable()
+		if err != nil {
+			klog.Error("failed to generate mounts table.")
+			mp.destination <- defaultMsg
+		}
+		mtDiff := libmount.GenerateDiff(mp.mountTable, newMountTable)
+		mp.mountTable = newMountTable
+		mp.processDiff(mtDiff)
+	}
+}
+
+func (mp *mountProbe) newMountTable() (*libmount.MountTab, error) {
+	return libmount.NewMountTab(libmount.FromFile(mp.mountsFileName,
+		libmount.MNT_FMT_FSTAB),
+		libmount.WithAllowFilter(libmount.SourceContainsFilter("/dev/")),
+		libmount.WithDenyFilter(libmount.SourceFilter("overlay")),
+		libmount.WithDenyFilter(libmount.TargetContainsFilter("/var/lib/kubelet/pod")))
+}
+
+func (mp *mountProbe) processDiff(diff libmount.MountTabDiff) {
+	changedDevices := make(map[string]struct{})
```

Comment:
Why is this empty struct needed here?

Reply:
I need the functionality of a set here. A single diff may have multiple changes related to a device. For example, say a device is mounted at some point A. Now it can be moved to B and also be mounted at a new point C. These are two different changes, but the underlying device is the same. Hence we would want to only capture unique occurrences of the devices.

Reply:
got it. But still why do we need an empty struct{}, that is not used.

Reply:
oh, that is to satisfy the golang requirements. we need to specify a type for the value in the map right. So I just use an empty struct since we don't really care about the value and only see if the key is set